common/formatting/layout_optimizer.cc

```diff
@@ -30,6 +32,103 @@
 
 namespace verible {
 
+void OptimizeTokenPartitionTree(const BasicFormatStyle& style,
+                                TokenPartitionTree* node,
+                                std::vector<PreFormatToken>* ftokens) {
+  CHECK_NOTNULL(node);
+
+  VLOG(4) << __FUNCTION__ << ", before:\n" << *node;
+  const auto indentation = node->Value().IndentationSpaces();
+
+  LayoutFunctionFactory factory(style);
+
+  const std::function<LayoutFunction(const TokenPartitionTree&)> TraverseTree =
+      [&TraverseTree, &style, &factory](const TokenPartitionTree& subnode) {
+        const auto policy = subnode.Value().PartitionPolicy();
+
+        if (subnode.is_leaf()) {
+          return factory.Line(subnode.Value());
+        }
+
+        switch (policy) {
+          case PartitionPolicyEnum::kOptimalFunctionCallLayout: {
+            // Support only function/macro/system calls for now
+            CHECK_EQ(subnode.Children().size(), 2);
+
+            const auto& function_header = subnode.Children()[0];
+            const auto& function_args = subnode.Children()[1];
+
+            auto header = TraverseTree(function_header);
+            auto args = TraverseTree(function_args);
+
+            auto stack_layout = factory.Stack({
+                header,
+                factory.Indent(args, style.wrap_spaces),
+            });
+            if (args.MustWrap()) {
+              return stack_layout;
+            }
+            auto juxtaposed_layout = factory.Juxtaposition({
+                header,
+                args,
+            });
+            return factory.Choice({
+                std::move(juxtaposed_layout),
+                std::move(stack_layout),
+            });
+          }
+
+          case PartitionPolicyEnum::kAppendFittingSubPartitions:
+          case PartitionPolicyEnum::kFitOnLineElseExpand: {
+            absl::FixedArray<LayoutFunction> layouts(subnode.Children().size());
+            std::transform(subnode.Children().begin(), subnode.Children().end(),
+                           layouts.begin(),
+                           [&TraverseTree](const TokenPartitionTree& subnode) {
+                             return TraverseTree(subnode);
+                           });
```

Comment:
Can this lambda be simplified to just passing `TraverseTree`?  (same question below on line 97)

Reply:
Fixed.